File: 177_RottenOranges.cpp
/*
Problem: Rotten Oranges
Description: Find minimum time required to rot all oranges. Fresh orange rots if adjacent to rotten orange.

Approach: BFS from all initially rotten oranges. Track time level-wise.
*/

#include <bits/stdc++.h>
using namespace std;

int orangesRotting(vector<vector<int>>& grid) {
    int rows = grid.size(), cols = grid[0].size();
    queue<pair<int, int>> q;
    int fresh = 0;
    
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            if (grid[i][j] == 2) {
                q.push({i, j});
            } else if (grid[i][j] == 1) {
                fresh++;
            }
        }
    }
    
    if (fresh == 0) return 0;
    
    int time = 0;
    int dx[] = {-1, 1, 0, 0};
    int dy[] = {0, 0, -1, 1};
    
    while (!q.empty()) {
        int size = q.size();
        bool rotted = false;
        
        for (int i = 0; i < size; i++) {
            auto [x, y] = q.front();
            q.pop();
            
            for (int d = 0; d < 4; d++) {
                int nx = x + dx[d];
                int ny = y + dy[d];
                
                if (nx >= 0 && nx < rows && ny >= 0 && ny < cols && grid[nx][ny] == 1) {
                    grid[nx][ny] = 2;
                    q.push({nx, ny});
                    fresh--;
                    rotted = true;
                }
            }
        }
        
        if (rotted) time++;
    }
    
    return fresh == 0 ? time : -1;
}

int main() {
    vector<vector<int>> grid = {
        {2, 1, 1},
        {1, 1, 0},
        {0, 1, 1}
    };
    
    cout << "Time to rot all oranges: " << orangesRotting(grid) << endl;
    // Output: 4
    
    return 0;
}

/*
Time Complexity: O(rows * cols)
Space Complexity: O(rows * cols)
*/

File: 177_RottenOranges.py
"""
Problem: Rotten Oranges
Description: Find minimum time required to rot all oranges. Fresh orange rots if adjacent to rotten orange.

Approach: BFS from all initially rotten oranges. Track time level-wise.
"""

from typing import List
from collections import deque

def orangesRotting(grid: List[List[int]]) -> int:
    rows, cols = len(grid), len(grid[0])
    queue = deque()
    fresh = 0
    
    for i in range(rows):
        for j in range(cols):
            if grid[i][j] == 2:
                queue.append((i, j))
            elif grid[i][j] == 1:
                fresh += 1
    
    if fresh == 0:
        return 0
    
    time = 0
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    
    while queue:
        size = len(queue)
        rotted = False
        
        for _ in range(size):
            x, y = queue.popleft()
            
            for dx, dy in directions:
                nx, ny = x + dx, y + dy
                
                if 0 <= nx < rows and 0 <= ny < cols and grid[nx][ny] == 1:
                    grid[nx][ny] = 2
                    queue.append((nx, ny))
                    fresh -= 1
                    rotted = True
        
        if rotted:
            time += 1
    
    return time if fresh == 0 else -1

def main():
    grid = [
        [2, 1, 1],
        [1, 1, 0],
        [0, 1, 1]
    ]
    
    print(f"Time to rot all oranges: {orangesRotting(grid)}")
    # Output: 4

if __name__ == "__main__":
    main()

"""
Time Complexity: O(rows * cols)
Space Complexity: O(rows * cols)
"""
